src/Streamly/Internal/Data/Stream/Channel/Type.hs

```diff
@@ -820,13 +826,30 @@ dumpChannel sv = do
 -- Cleanup
 -------------------------------------------------------------------------------
 
--- | Never called from a worker thread.
-cleanupChan :: Channel m a -> IO ()
-cleanupChan chan = do
-    stopped <- atomicModifyIORef (channelStopped chan) (\x -> (True,x))
+channelDone :: Channel m a -> String -> IO ()
+channelDone chan reason = do
+    when (svarInspectMode chan) $ do
+        t <- getTime Monotonic
+        writeIORef (svarStopTime (svarStats chan)) (Just t)
+        printSVar (dumpChannel chan) reason
+        -- If there are any other channels referenced by this channel a GC will
+        -- prompt them to be cleaned up quickly.
+        performMajorGC
+
+-- XXX Handle ThreadAbort while this is running?
+
+-- | Never called from a worker thread. This can be called multiple times and
+-- only one will succeed others will block until it finishes.
+--
+cleanupChan :: Channel m a -> String -> IO ()
+cleanupChan chan reason = do
```

Comment:
There will always be a major GC performed before the channel stops.
It may not be required in many cases where there isn't any resource allocation.
Is it possible to `performMajorGC` only when required?

Reply:
performMajorGC is done only in inspect mode so that logs come properly.